### PR TITLE
Add initial project generation infrastructure

### LINKED ISSUE #14: Figure out project generation...
I have the Generator GUI and @Piphi5 has some generation code sitting around. We need to decide whether we should merge this or create another avenue for teams to log data (i.e. API in WPILib).

### LINKED ISSUE #37: Add Romi generation preset
frc-char requires special setup for characterizing a Romi. While constants are mostly plug-n-play for users of Romi, it's a great learning platform, and we should add a preset for Romi.

### LINKED ISSUE #41: Refactor NT logging in integration_test_project into SysIdLogger classes
Even if we do code generation, the abstraction there will be helpful in codegenned projects as well as teams who decide to gather characterization data in situ anyway.

### PR DESCRIPTION
**Pull request by Piphi5**
TODO:
Fixes #14 
- [x] Finalize Generator GUI
- [x] Finish Single-Sided Motor Project
- [x] Finish Drive Project
- [x] Create Tests to ensure the validity of these projects
- [x] Fix #37 By adding Romi Preset
- [x] Update data to match #38 
- [x] Add the proper CPR conversions for motor controllers
- [x] Fix #41 by adding SysId Classes
- [x] Fix widget sizing
- [x] ~~Set CAN update rates to 5 ms (need to figure out CAN issues)~~ (File an issue, and resolve later so that we can get the fundamental generator merged + tested sooner)
- [x] Have settings for timing and average for different types of encoders
- [x] ~~Add venom support~~ (File an issue, and resolve later so that we can get the fundamental generator merged + tested sooner)
- [x] Optimize the logging by auto calculating the voltage
- [x] Update Telemetry Docs
- [x] Fix linter errors
- [x] Update Usage Docs in README

### CODE REVIEW

**Comment by Daltz333** [comment]
Adapt initial analyzer/generator/logger screens to fit application start width? I had to resize widgets because they were huge.

**Comment by Daltz333** [comment]
Also, the plus/minus here should really be more obvious buttons
image

**Comment by Daltz333** [comment]
For drivetrain control, it really should be per side, teams can quite easily destroy their gearbox by accidentally unchecking one side inversion.

**Comment by Piphi5** [comment]
How should I make it more obvious? Should I put it on a separate line?

**Comment by Daltz333** [comment]
If you could make it look more like a button but with the text + or -

**Comment by Piphi5** [comment]
image
Is this setup better?

**Comment by Daltz333** [comment]
Yes

**Comment by Piphi5** [comment]
Romi Gradle changes for future reference:
```
dependencies {
    //simulation wpi.deps.sim.gui(wpi.platforms.desktop, true)
    
    // Enable simulation gui support. Must check the box in vscode to enable support
    // upon debugging
    simulation wpi.deps.sim.gui(wpi.platforms.desktop, true)
    simulation wpi.deps.sim.driverstation(wpi.platforms.desktop, true)

    // Websocket extensions require additional configuration.
    // simulation wpi.deps.sim.ws_server(wpi.platforms.desktop, false)
    simulation wpi.deps.sim.ws_client(wpi.platforms.desktop, true)

    // Websocket extensions require additional configuration.
    // simulation wpi.deps.sim.ws_server(wpi.platforms.desktop, true)
    // simulation wpi.deps.sim.ws_client(wpi.platforms.desktop, true)
}

// Set the websocket remote host (the Romi IP address).
sim {
  envVar "HALSIMWS_HOST", "10.0.0.2"
}
```

**Comment by Piphi5** [comment]
image
This is the GUI as of right now if anyone has any changes they want to request.

File: src/integrationtest/native/cpp/GenerationTest.cpp [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+const wpi::SmallVector<std::string, 4> kPigeonCtors{"0", "WPI_TalonSRX-1"};
+
+const wpi::SmallVector<std::string, 4> kAnalogCtors{"0"};
+
+const wpi::SmallVector<std::string, 4> kNavXCtors{"SPI.kMXP"};
+
+const wpi::SmallVector<std::string, 4> kADXRS450Ctors{"SPI.kMXP"};
+
+std::map<std::string, wpi::SmallVector<std::string, 4>> gyroCtorMap = {
```
Comment (prateekma) [inline src/integrationtest/native/cpp/GenerationTest.cpp:34]: Can we use `wpi::StringMap` here?

File: src/main/native/cpp/generation/ConfigManager.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+#include <stdexcept>
+
+#include <wpi/json.h>
+#include <wpi/raw_ostream.h>
+
+using namespace sysid;
+
+ConfigManager::ConfigManager(ConfigSettings& config) : m_config(config) {}
+
+template <class T>
```
Comment (prateekma) [inline src/main/native/cpp/generation/ConfigManager.cpp:17]: Use `typename` for consistency.

Comment (Piphi5) [reply src/main/native/cpp/generation/ConfigManager.cpp:17]: instead of class?

Comment (prateekma) [reply src/main/native/cpp/generation/ConfigManager.cpp:17]: Yes

File: base_projects/Drivetrain/src/main/cpp/Robot.cpp [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+#include <frc/smartdashboard/SmartDashboard.h>
+#include <units/angular_velocity.h>
+#include <wpi/FileSystem.h>
+#include <wpi/SmallString.h>
+#include <wpi/StringMap.h>
+
+// #include "AHRS.h"
+#include "rev/CANEncoder.h"
+
+Robot::Robot() : frc::TimedRobot(5_ms) {
```
Comment (calcmogul) [inline base_projects/Drivetrain/src/main/cpp/Robot.cpp:29]: To improve scheduling jitter, we should also set the HAL Notifier thread and the TimedRobot thread to RT, like this: https://github.com/frc3512/Robot-2020/blob/main/src/main/cpp/Robot.cpp#L87-L93. 3512 uses a priority of 15.

File: base_projects/Drivetrain/src/main/cpp/Robot.cpp [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+
+  } else {
+    frc::filesystem::GetDeployDirectory(path);
+    os << "/config.json";
+  }
+
+  wpi::raw_fd_istream is{path.c_str(), ec};
+
+  if (ec) {
+    std::cout << "File error: " << path.c_str() << std::endl;
```
Comment (calcmogul) [inline base_projects/Drivetrain/src/main/cpp/Robot.cpp:46]: This applies to other instances of std::cout.
```suggestion
    wpi::outs() << "File error: " << path.c_str() << '\n';
    wpi::outs().flush();
```

File: base_projects/Drivetrain/src/main/cpp/Robot.cpp [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+        port_str = std::string{gyroCtor[gyroCtor.find("-") + 1]};
+      } else {
+        port_str = gyroCtor;
+      }
+      // converts gyroCtor into port #
+      std::stringstream ss(port_str);
+      int srx_port = 0;
+      ss >> srx_port;
+      if (gyroCtor.contains("WPI_TalonSRX")) {
+        m_pigeon = std::make_unique<PigeonIMU>(new WPI_TalonSRX(srx_port));
```
Comment (calcmogul) [inline base_projects/Drivetrain/src/main/cpp/Robot.cpp:109]: Does the PigeonIMU class actually take a TalonSRX in its ctor?

File: base_projects/Drivetrain/src/main/cpp/Robot.cpp [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+        // } else if (gyroType == "NavX") {
+        //   if (gyroCtor == "SerialPort.kUSB") {
+        //     m_gyro = std::make_unique<AHRS>(frc::SerialPort::Port::kUSB);
+        //   } else if (gyroCtor == "I2C") {
+        //     m_gyro = std::make_unique<AHRS>(frc::I2C::Port::kMXP);
+        //   } else if (gyroCtor == "SerialPort.kMXP") {
+        //     m_gyro = std::make_unique<AHRS>(frc::SerialPort::Port::kMXP);
+        //   } else {
+        //     m_gyro = std::make_unique<AHRS>(frc::SPI::Port::kMXP);
+        //   }
```
Comment (calcmogul) [inline base_projects/Drivetrain/src/main/cpp/Robot.cpp:145]: Should this be removed?

File: base_projects/Drivetrain/src/main/cpp/Robot.cpp [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+    }
+    static_cast<rev::CANSparkMax*>(controllers->back().get())
+        ->SetInverted(inverted);
+  } else {
+    controllers->emplace_back(std::make_unique<frc::Spark>(port));
+    static_cast<frc::Spark*>(controllers->back().get())->SetInverted(inverted);
+  }
+}
+
+void Robot::SetupEncoders(const std::string& controllerName,
```
Comment (calcmogul) [inline base_projects/Drivetrain/src/main/cpp/Robot.cpp:208]: I think this is the right thing? It's either `const wpi::Twine&` or `wpi::StringRef.
```suggestion
void Robot::SetupEncoders(const wpi::Twine& controllerName,
```

File: base_projects/Drivetrain/src/main/cpp/Robot.cpp [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+          ->SetSensorPhase(encoderInverted);
+      position = [&, this] {
+        return dynamic_cast<WPI_BaseMotorController*>(controller)
+                   ->GetSelectedSensorPosition() /
+               m_cpr;
+      };
+      rate = [&, this] {
+        return dynamic_cast<WPI_BaseMotorController*>(controller)
+                   ->GetSelectedSensorVelocity() /
+               m_cpr / .1;
```
Comment (calcmogul) [inline base_projects/Drivetrain/src/main/cpp/Robot.cpp:236]: This magic number should be a constant or documented somehow.
```suggestion
               m_cpr / 0.1;
```

File: base_projects/Drivetrain/src/main/include/logging/SysIdDrivetrainLogger.h [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  /**
+   * The users should set their right motors to what this returns AFTER calling
+   * log.
+   */
+  units::volt_t GetRightMotorVoltage();
+
+  /**
+   * Logs data for a drivetrain mechanism.
+   * Outputs data in form: timestamp, l voltage, r voltage, l position, r
+   * position, l velocity, r velocity, angle, angular rate
```
Comment (calcmogul) [inline base_projects/Drivetrain/src/main/include/logging/SysIdDrivetrainLogger.h:26]: This should be more like
```suggestion
   * Logs data for a drivetrain mechanism.
   *
   * Outputs data in form: timestamp, l voltage, r voltage, l position, r
   * position, l velocity, r velocity, angle, angular rate.
   *
   * @param leftPosition Something descriptive.
   * @param rightPosition Something descriptive.
   * <you get the idea>
```

File: base_projects/Drivetrain/src/main/include/logging/SysIdLogger.h [hunk 10]
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <sstream>
```
Comment (calcmogul) [inline base_projects/Drivetrain/src/main/include/logging/SysIdLogger.h:9]: These aren't used in this header; remove them. If they are used in SysIdLogger.cpp, they should be moved there.

**Review by calcmogul — CHANGES_REQUESTED** [review]

File: base_projects/Drivetrain/src/main/cpp/Robot.cpp (hunk 3 above)
Comment (Piphi5) [reply base_projects/Drivetrain/src/main/cpp/Robot.cpp:29]: Should this be something we might abstract into the Logging API?

File: base_projects/Drivetrain/src/main/cpp/Robot.cpp (hunk 6 above)
Comment (Piphi5) [reply base_projects/Drivetrain/src/main/cpp/Robot.cpp:145]: Maybe for now? The issue is that NavX doesn't have Linux or mac binaries so the integration tests fail for it, but ideally we should be able to support it for the real version.

Comment (calcmogul) [reply base_projects/Drivetrain/src/main/cpp/Robot.cpp:145]: I guess we could mark it as `FIXME: waiting on Linux and macOS builds for navX AHRS` to save us work later.

File: base_projects/Drivetrain/src/main/cpp/Robot.cpp (hunk 3 above)
Comment (calcmogul) [reply base_projects/Drivetrain/src/main/cpp/Robot.cpp:29]: Abstract in what way?

File: base_projects/Drivetrain/src/main/cpp/Robot.cpp (hunk 8 above)
Comment (Piphi5) [reply base_projects/Drivetrain/src/main/cpp/Robot.cpp:236]: Would it be better to use the units library to specify this is converting the rawSensorUnits/100ms into rotations/second?

File: base_projects/Drivetrain/src/main/cpp/Robot.cpp (hunk 3 above)
Comment (Piphi5) [reply base_projects/Drivetrain/src/main/cpp/Robot.cpp:30]: Is [this](https://github.com/Piphi5/sysid/blob/generation/integration_test_project/src/main/cpp/logging/SysIdLogger.cpp#L41-#L48) a good spot to put it?

**Comment by prateekma** [comment]
The default window positions cause the text in the Logger to be cut off:
image

**Comment by Daltz333** [comment]
1. Missing necessary ``build.gradle`` changes for the Romi
2. Robot Program crashes when a test is ran

image

**Comment by Piphi5** [comment]
@prateekma  The Logger Text cutoff should be fixed now:
image


@Daltz333  The build.gradle changes have been implemented now. When does the romi crash? Is it after the test is disabled? If so, could you send the telemetry string that the project spits out?

**Comment by Daltz333** [comment]
It crashes immediately when I press autonomous in sim gui

**Comment by Piphi5** [comment]
@Daltz333 could you try deleting your clone and trying again?

File: src/integrationtest/native/cpp/GenerationTest.cpp [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+#include "gtest/gtest.h"
+#include "networktables/NetworkTableValue.h"
+#include "sysid/Util.h"
+#include "sysid/generation/ConfigManager.h"
+#include "sysid/view/Generator.h"
+
+using namespace std::chrono_literals;
+
+const wpi::SmallVector<std::string, 4> kPigeonCtors{"0", "WPI_TalonSRX-1"};
+
```
Comment (prateekma) [inline src/integrationtest/native/cpp/GenerationTest.cpp:31]: Remove spaces between these

File: src/integrationtest/native/cpp/GenerationTest.cpp [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+    {"AnalogGyro", kAnalogCtors},
+    {"Pigeon", kPigeonCtors},
+    {"ADXRS450", kADXRS450Ctors},
+    {"NavX", kNavXCtors}};
+
+class GenerationTest : public ::testing::Test {
+ public:
+  GenerationTest() : m_manager(m_settings) {}
+
+  void SetUp(std::string directory) {
```
Comment (prateekma) [inline src/integrationtest/native/cpp/GenerationTest.cpp:48]: use wpi::StringRef

**Review by prateekma — CHANGES_REQUESTED** [review]

**Comment by prateekma** [comment]
We should rename the "cpr" entry in the JSON to something else to not cause confusion when human-reading it.

**Comment by Piphi5** [comment]
does "counts per shaft revolution" work?

**Comment by Piphi5** [comment]
Current setup:
image

**Comment by calcmogul** [comment]
> Set CAN update rates to 5 ms (need to figure out CAN issues)

If we can get clang-tidy passing and there's no regressions in this PR as is, we could file an issue to figure out 5ms sampling later. Consistent sampling is more important than high frequency sampling, and the only systems that need 5ms sampling are really quickly accelerating things like flywheels.

**Comment by Piphi5** [comment]
Yeah we could do that.

The clang-tidy thing is a problem that I have no idea how to solve though.

**Comment by calcmogul** [comment]
My guess is the define not being properly escaped with quotation marks, but I don't know how to make gradle generate just a specific preprocessed file (with gcc's -E flag) to confirm that.

**Comment by Piphi5** [comment]
Also, the normal `build.gradle` in the robot projects, uses the same define method, so it's specific to how this gradle file is setup.

**Comment by calcmogul** [comment]
I made a PR to this branch fixing the stringification.

**Comment by calcmogul** [comment]
It says the macOS integration test for arms failed with an exception.

File: README.md [hunk 13]
```diff
@@ -27,6 +27,17 @@ The SysId GitHub repository uses GitHub Actions to build and test each commit.
...
 5. Run the executable named `sysid`.
 
+## Downloading and Running Full Development Versions of SysId (Generation, Logging, Analysis)
+
+Clone the SysId Github repository to get all the necessary code.
+
+Go to the directory that the repository is cloned in and run `./gradlew run` to start the executable.
+
+### Generating And Running SysId Robot Code
+1. Use the Generator Widget to create a config.json
```
Comment (calcmogul) [inline README.md:37]: Add a newline between the heading and bulleted list like the others.

File: README.md [hunk 14]
```diff
@@ -46,3 +57,11 @@ There is also a robot project in `integration_test_project` that you can use to
     - On Linux, install GCC 7 or greater
     - On Windows, install [Visual Studio Community 2019](https://visualstudio.microsoft.com/vs/community/) and select the C++ programming language during installation (Gradle can't use the build tools for Visual Studio 2019)
     - On macOS, install the Xcode command-line build tools via `xcode-select --install`
+
+## Logging Projects
```
Comment (calcmogul) [inline README.md:61]: Add a newline between the heading and the paragraph.

File: README.md [hunk 15]
```diff
@@ -46,3 +57,11 @@ There is also a robot project in `integration_test_project` that you can use to
...
+SysId comes with projects that interface with the telemetry manager to provide the necessary data for analysis. These projects are stored in the `base_projects` folder and take in a `config.json` file in the `src/main/deploy` directory to setup the robot hardware for analysis.
+
+There is a `Drivetrain` project for drivetrain analysis and a `GeneralMechanism` project for simple-motor, elevator, and arm analyses.
+
+### Running the Projects
```
Comment (calcmogul) [inline README.md:66]: Add a newline between the heading and the paragraph.

**Review by calcmogul — CHANGES_REQUESTED** [review]

**Review by calcmogul — DISMISSED** [review]

File: docs/data-collection.md [hunk 16]
```diff
@@ -9,7 +9,8 @@ Here is a list of the NT entries that are used to send and collect data between
 |             NT Entry             |   Type   |                                                                    Description                                                                     |
 | -------------------------------- | -------- | -------------------------------------------------------------------------------------------------------------------------------------------------- |
 | `/SmartDashboard/SysIdTelemetry` | `string` | Used to send telemetry from the robot program. This data is sent after the test completes once the robot enters the disabled state.                |
-| `/SmartDashboard/SysIdAutoSpeed` | `double` | Used to receive the latest autospeed value from the Logger.                                                                                        |
+| `/SmartDashboard/SysIdVoltageCommand` | `double` | Used to either send the ramp rate (V/s) for the quasistatic test or the voltage (V) for the dynamic test.                                                                                        |
```
Comment (prateekma) [inline docs/data-collection.md:12]: Can you line up the `|` on the table?

File: integration_test_project/src/main/cpp/Robot.cpp [hunk 17]
```diff
@@ -23,40 +24,32 @@
...
   }
 
   void DisabledInit() override {
-    m_mechanism->SetMotor(0);
+    m_mechanism->SetMotor(units::volt_t{0});
 
-    m_driveMechanism->SetLMotor(0);
-    m_driveMechanism->SetRMotor(0);
+    m_driveMechanism->SetLMotor(units::volt_t{0});
+    m_driveMechanism->SetRMotor(units::volt_t{0});
```
Comment (prateekma) [inline integration_test_project/src/main/cpp/Robot.cpp:44]: Use `0_V`.

File: integration_test_project/src/main/cpp/Robot.cpp [hunk 18]
```diff
@@ -23,40 +24,32 @@
...
-
-      frc::SmartDashboard::PutString("SysIdTelemetry", ss.str());
-
-      m_counter = 0;
-      m_data.clear();
+    if (m_test == "Drivetrain") {
+      std::cout << "SEND DRIVE\n";
+      m_driveLogger.SendData();
+    } else {
+      std::cout << "SEND Gneral\n";
```
Comment (prateekma) [inline integration_test_project/src/main/cpp/Robot.cpp:51]: Typo (and did you mean to capitalize the entire word like in drivetrain above?)

File: integration_test_project/src/main/cpp/Robot.cpp [hunk 19]
```diff
@@ -69,11 +62,18 @@ class Robot : public frc::TimedRobot {
   }
 
   void AutonomousInit() override {
+    m_test = frc::SmartDashboard::GetString("SysIdTest", "Simple");
+    if (m_test == "Drivetrain") {
+      std::cout << "DRIVE\n";
+      m_driveLogger.InitLogging();
+    } else {
+      std::cout << "General\n";
```
Comment (prateekma) [inline integration_test_project/src/main/cpp/Robot.cpp:70]: See comment wrt capitalization above

**Review by prateekma — CHANGES_REQUESTED** [review]

File: integration_test_project/src/main/cpp/Robot.cpp (hunk 18 above)
Comment (Piphi5) [reply integration_test_project/src/main/cpp/Robot.cpp:51]: Oops, those were mainly debug prints

**Review by prateekma — APPROVED** [review]

**Review by calcmogul — DISMISSED** [review]

**Comment by Piphi5** [comment]
I just rebased and squashed a commit

**Review by prateekma — APPROVED** [review]

**Pull request merged**